### PR TITLE
Tests for CIS 194 homework assignments (1 to 6)

### PR DESCRIPTION
**Pull request by ghost**
I have added tests based in homeworks (1 to 6) of CIS 194 course. The reason for this pull request is that all tests that use `doctest` package exists. The `doctest` package definition is:

> The doctest program checks examples in source code comments. It is modeled after doctest for Python.

It can cause confusion at the time of run tests, because usually tests source code are in other files and folders. I think that a better practice is to use a testing package for this tests. I choosed `hspec` package because I found this package in some Stack Builders repositories. The `hspec` package definition is:

> Hspec is a testing framework for Haskell. It is inspired by the Ruby library RSpec.

 I have used `hspec` package to migrate `doctest` tests to this testing framework.

Please check these tests and write any commentary if is necessary.

### CODE REVIEW

File: homework01/test/examples/Main.hs [hunk 1]
```diff
@@ -1,15 +1,44 @@
-module Main
-  ( main
```
Comment (juanpaucar) [inline homework01/test/examples/Main.hs]: I think we should keep this to be explicit about what the module exports

File: homework01/test/examples/Main.hs [hunk 2]
```diff
@@ -1,15 +1,44 @@
...
 
 main :: IO ()
-main =
-  DocTest.doctest
-    [ "-isrc"
-    , "src/Basis.hs"
-    ]
+main = hspec $ do
+
+  describe "toDigits" $ do
```
Comment (juanpaucar) [inline homework01/test/examples/Main.hs:11]: Not sure if it's important, but what about the case for numbers starting with `0` like `01234`

File: homework01/test/examples/Main.hs [hunk 3]
```diff
@@ -1,15 +1,44 @@
...
+    it "converts positive integers to a list of digits" $
+      toDigits 1234 `shouldBe` [1, 2, 3, 4]
+
+    it "returns an empty list for 0" $
+      toDigits 0 `shouldBe` []
+
+    it "returns an empty list for negative inputs" $
+      toDigits (-17) `shouldBe` []
+
+  describe "toDigitsRev" $
```
Comment (juanpaucar) [inline homework01/test/examples/Main.hs]: What does this function makes if a negative number is passed as argument? and like before, not sure if important but what if the number starts with `0`

File: homework01/test/examples/Main.hs [hunk 4]
```diff
@@ -1,15 +1,44 @@
...
+    it "converts positive integers to a list of digits reversed" $
+      toDigitsRev 1234 `shouldBe` [4, 3, 2, 1]
+
+  describe "doubleEveryOther" $ do
```
Comment (juanpaucar) [inline homework01/test/examples/Main.hs]: Not sure if important, but what happens with an empty list?

File: homework02/test/examples/Main.hs [hunk 5]
```diff
@@ -1,15 +1,39 @@
-module Main
-  ( main
```
Comment (juanpaucar) [inline homework02/test/examples/Main.hs]: Again, I think it's a better idea to keep specific what the module exports

File: homework01/test/examples/Main.hs [hunk 6]
```diff
@@ -1,15 +1,44 @@
...
+  describe "validate" $ do
+    it "returns True if an Integer is a valid credit card number" $
+      validate 4012888888881881 `shouldBe` True
+    
+    it "returns True if an Integer is a invalid credit card number" $
+      validate 4012888888881882 `shouldBe` False
+
+  describe "hanoi" $
+    it "gets a list of moves of the stack of discs from first peg to second" $
+      hanoi 2 "a" "b" "c" `shouldBe` [("a", "c"), ("a", "b"), ("c", "b")]
```
Comment (juanpaucar) [inline homework01/test/examples/Main.hs]: I think it would be better to increase the disc number to check it works to output more  complex solutions

File: homework02/test/examples/Main.hs [hunk 7]
```diff
@@ -1,15 +1,39 @@
...
+    it "parses an info message" $
+      parseMessage "I 29 la la la" `shouldBe` LogMessage Info 29 "la la la"
+
+    it "parses an Unknown message" $
+      parseMessage "This is not in the right format" `shouldBe` Unknown "This is not in the right format"  
+
+  describe "insert" $
+    it "returns a tree with LogMessage inserted" $ do
+      let tree = Node Leaf (LogMessage (Error 2) 562 "help help") Leaf
+      insert (LogMessage Info 29 "la") tree `shouldBe` (Node (Node Leaf (LogMessage Info 29 "la") Leaf) (LogMessage (Error 2) 562 "help help") Leaf)
```
Comment (juanpaucar) [inline homework02/test/examples/Main.hs]: could you split this in more than one line to make it more readable

File: homework02/test/examples/Main.hs [hunk 8]
```diff
@@ -1,15 +1,39 @@
...
+   
+  describe "build" $ 
+    it "builds up a MessageTree containing the messages in a list" $ do
+      let messageTreeSet = [LogMessage (Error 2) 562 "help help", LogMessage Info 29 "la"]
+      build messageTreeSet `shouldBe` Node (Node Leaf (LogMessage Info 29 "la") Leaf) (LogMessage (Error 2) 562 "help help") Leaf
```
Comment (juanpaucar) [inline homework02/test/examples/Main.hs]: the same as above, please split this into multiple lines to make it more readable

File: homework02/test/examples/Main.hs [hunk 9]
```diff
@@ -1,15 +1,39 @@
...
+  
+  describe "inOrder" $
+    it "produces a list of all LogMessages it contains, sorted by timestamp from smallest to biggest" $ do
+      let messageTreeSet = [LogMessage (Error 2) 562 "help help",LogMessage Info 29 "la"]
+      inOrder (build messageTreeSet) `shouldBe` [LogMessage Info 29 "la",LogMessage (Error 2) 562 "help help"]
+  
+  describe "whatWentWrong" $
+    it "returns a list of the messages corresponding to any errors with a severity of 50 or greater, sorted by timestamp" $
```
Comment (juanpaucar) [inline homework02/test/examples/Main.hs]: This is a long description, probably it's a good a idea to put it inside a `context` say for example "severity less than 50" and another context "severity of 50 or greater"

File: homework04/test/examples/Main.hs [hunk 10]
```diff
@@ -1,15 +1,37 @@
-module Main
-  ( main
```
Comment (juanpaucar) [inline homework04/test/examples/Main.hs]: same as above, let's keep being specific about what the module exports

File: homework04/test/examples/Main.hs [hunk 11]
```diff
@@ -1,15 +1,37 @@
...
+      (height $ foldTree "ABCDEFGHIJ") `shouldBe` 3
+         
+  describe "xor" $ do
+    it "returns xor from a list with an odd number of True values" $
+      xor [False, True, False] `shouldBe` True
+   
+    it "returns xor from a list with an even number of True values" $
+      xor [False, True, False, False, True] `shouldBe` False
+
+  describe "map'" $
```
Comment (juanpaucar) [inline homework04/test/examples/Main.hs]: what about an empty list?

File: homework04/test/examples/Main.hs [hunk 12]
```diff
@@ -1,15 +1,37 @@
...
+    it "returns same output of standard map function" $
+      map' (+1) [1,2,3] `shouldBe` map (+1) [1,2,3]
+  
+  describe "sieveSundaram" $
+    it "returns a list of prime numbers up to 2*n + 2" $
```
Comment (juanpaucar) [inline homework04/test/examples/Main.hs]: although not that important, what happens with negative numbers passed?

File: homework05/test/examples/Main.hs [hunk 13]
```diff
@@ -1,15 +1,61 @@
-module Main
-  ( main
```
Comment (juanpaucar) [inline homework05/test/examples/Main.hs]: same as above, let's keep being specific about what a module exports

File: homework05/test/examples/Main.hs [hunk 14]
```diff
@@ -1,15 +1,61 @@
...
+      eval (ExprT.Mul (ExprT.Add (ExprT.Lit 2) (ExprT.Lit 3)) (ExprT.Lit 4)) `shouldBe` 20
+
+  describe "evalStr" $ do
+    it "evaluates expression with parenthesis" $
+      evalStr "(2+3)*4" `shouldBe` Just 20
+
+    it "evaluates expression without parenthesis" $
+      evalStr "2+3*4" `shouldBe` Just 14
+
+    it "evaluates wrong expression" $
```
Comment (juanpaucar) [inline homework05/test/examples/Main.hs]: can we make a case for a negative number too?

File: homework05/test/examples/Main.hs [hunk 15]
```diff
@@ -1,15 +1,61 @@
...
+    
+      it "returns value of the expression evaluated as MinMax" $ do
+        let testMM = parseExp lit add mul "(3 * 4) + (-5)" :: Maybe MinMax
+        testMM `shouldBe` Just (MinMax 3)
+    
+      it "returns value of the expression evaluated as Mod7" $ do
+        let testSat = parseExp lit add mul "(3 * 4) + (-5)" :: Maybe Mod7
+        testSat `shouldBe` Just (Mod7 0)
+   
+  describe "compile" $
```
Comment (juanpaucar) [inline homework05/test/examples/Main.hs]: what happens with incorrect input?

File: homework06/test/examples/Main.hs [hunk 16]
```diff
@@ -1,15 +1,45 @@
-module Main
-  ( main
```
Comment (juanpaucar) [inline homework06/test/examples/Main.hs]: same as above about what the module exports

File: homework05/test/examples/Main.hs [hunk 17]
```diff
@@ -1,15 +1,61 @@
...
+
+    it "evaluates expression without parenthesis" $
+      evalStr "2+3*4" `shouldBe` Just 14
+
+    it "evaluates wrong expression" $
+      evalStr "2+3*" `shouldBe` Nothing
+      
+  describe "reify" $
+    it "makes an instance of Expr to ExprT type" $
+      reify (mul (add (lit 2) (lit 3)) (lit 4)) `shouldBe` ExprT.Mul (ExprT.Add (ExprT.Lit 2) (ExprT.Lit 3)) (ExprT.Lit 4)
```
Comment (juanpaucar) [inline homework05/test/examples/Main.hs]: please, split this line to make it more readable

**Review by juanpaucar — CHANGES_REQUESTED** [review]

File: homework02/test/examples/Main.hs [hunk 18]
```diff
@@ -1,15 +1,39 @@
...
+      pendingWith " whatWentWrong returns IO [String]"
+      -- testWhatWentWrong parse whatWentWrong "src/sample.log" `shouldBe` ["Way too many pickles","Bad pickle-flange interaction detected","Flange failed!"]
```
Comment (flandrade) [inline homework02/test/examples/Main.hs]: @jmorocho Could you please remove this comment? Thanks.

File: homework01/test/examples/Main.hs (hunk 1 above)
Comment (ghost) [reply homework01/test/examples/Main.hs]: Ok, no problem :+1:

File: homework01/test/examples/Main.hs (hunk 2 above)
Comment (ghost) [reply homework01/test/examples/Main.hs:11]: The signature of `toDigits` function is `toDigits :: Integer -> [Integer]`. Haskell don't use leading zeros, so `toDigits 01234 == toDigits 1234`.

File: homework01/test/examples/Main.hs (hunk 3 above)
Comment (ghost) [reply homework01/test/examples/Main.hs]: Checking homework 01, `toDigitsRev` function has the same tests as `toDigits` function, I'm going to add the corresponding tests.

File: homework01/test/examples/Main.hs (hunk 4 above)
Comment (ghost) [reply homework01/test/examples/Main.hs]: I'm not sure if it's necessary to add a test with an empty lists as input, because the exercise doesn't indicate this kind of input. I know that is necessary to test all cases (valid and invalid inputs) but this can produce confusion in a Haskell beginner when he test `doubleEveryOther` function and don't pass a test with an empty list.

File: homework01/test/examples/Main.hs (hunk 6 above)
Comment (ghost) [reply homework01/test/examples/Main.hs]: Ok, no problem :+1:

File: homework04/test/examples/Main.hs (hunk 11 above)
Comment (ghost) [reply homework04/test/examples/Main.hs]: `map'` function is defined by `foldr` function and control this case but you're right, original `map` function consider the behaviour when the second argument is an empty list. I'm going to add this test.

File: homework04/test/examples/Main.hs (hunk 12 above)
Comment (ghost) [reply homework04/test/examples/Main.hs]: In any implementation, you need to create a list as `[1...n]` in order to remove numbers of the form `i + j + 2ij` given some conditions, so if `n` is a negative number or zero, the output will be an empty list.

File: homework05/test/examples/Main.hs (hunk 14 above)
Comment (ghost) [reply homework05/test/examples/Main.hs]: I'm going to modify this tests to consider operations with negative numbers.

File: homework05/test/examples/Main.hs (hunk 15 above)
Comment (ghost) [reply homework05/test/examples/Main.hs]: `parseExp` function in `Parser.hs` file handles this case, producing `Nothing` for the inputs which are not well-formed expressions.

File: homework02/test/examples/Main.hs (hunk 7 above)
Comment (ghost) [reply homework02/test/examples/Main.hs]: Ok, no problem :+1:

File: homework05/test/examples/Main.hs (hunk 15 above)
Comment (juanpaucar) [reply homework05/test/examples/Main.hs]: ok

File: homework01/test/examples/Main.hs [hunk 19]
```diff
@@ -1,15 +1,56 @@
...
+      validate 4012888888881882 `shouldBe` False
+
+  describe "hanoi" $
+    context "when the stack discs moves from first peg to second" $ do
+      it "gets a list of moves whit 2 discs" $
+        hanoi 2 "a" "b" "c" `shouldBe` [("a", "c"), ("a", "b"), ("c", "b")]
+
+      it "gets a list of moves whit 3 discs" $
+        hanoi 3 "a" "b" "c" `shouldBe` [("a","b"),("a","c"),("b","c"),("a","b"),("c","a"),("c","b"),("a","b")]
+
```
Comment (flandrade) [inline homework01/test/examples/Main.hs]: @jmorocho Could you please add a EOF space?

File: homework02/test/examples/Main.hs [hunk 20]
```diff
@@ -1,15 +1,46 @@
...
+      let messageTreeSet = [LogMessage (Error 2) 562 "help help",LogMessage Info 29 "la"]
+      inOrder (build messageTreeSet) 
+        `shouldBe` [LogMessage Info 29 "la",LogMessage (Error 2) 562 "help help"]
+  
+  describe "whatWentWrong" $
+    context "with a severity of 50 or greater" $
+    it "returns a list of the messages corresponding to any errors sorted by timestamp" $
+      pendingWith "whatWentWrong returns IO [String]"
+      -- testWhatWentWrong parse whatWentWrong "src/sample.log" 
+          -- `shouldBe` ["Way too many pickles","Bad pickle-flange interaction detected","Flange failed!"]
```
Comment (flandrade) [inline homework02/test/examples/Main.hs]: @jmorocho Could you please remove this comment?

File: homework01/test/examples/Main.hs [hunk 21]
```diff
@@ -1,15 +1,56 @@
...
 main :: IO ()
-main =
-  DocTest.doctest
-    [ "-isrc"
-    , "src/Basis.hs"
-    ]
+main = hspec $ do
+
+  describe "toDigits" $ do
+    it "converts positive integers to a list of digits" $
```
Comment (flandrade) [inline homework01/test/examples/Main.hs]: @jmorocho Some of these describe a context. Could you please verify and add `context`? http://hspec.github.io/writing-specs.html

```haskell
describe "toDigits" $ do
    context "when digits are positive integers" $ do
      it "converts to a list of digits" $ do
```

**Review by flandrade — CHANGES_REQUESTED** [review]
@jmorocho Good job.  I left some comments. Could you please take a look? Thanks for your contribution!

In addition, could you please run `hlint`?

File: homework02/test/examples/Main.hs (hunk 18 above)
Comment (ghost) [reply homework02/test/examples/Main.hs]: Ok, no problem :+1:

File: homework02/test/examples/Main.hs (hunk 20 above)
Comment (ghost) [reply homework02/test/examples/Main.hs]: Ok, no problem :+1:

File: homework01/test/examples/Main.hs (hunk 21 above)
Comment (ghost) [reply homework01/test/examples/Main.hs]: You are right, the context is when digits are positive integers. I going to change this.

File: homework01/test/examples/Main.hs (hunk 19 above)
Comment (ghost) [reply homework01/test/examples/Main.hs]: Ok, no problem :+1:

File: homework02/test/examples/Main.hs (hunk 20 above)
Comment (ghost) [reply homework02/test/examples/Main.hs]: I have written the test of this function. Check please.

File: homework01/test/examples/Main.hs [hunk 22]
```diff
@@ -1,15 +1,57 @@
...
+    
+    it "doubles every other number beginning from the right (odd number of elements)" $
+      doubleEveryOther [1, 2, 3] `shouldBe` [1, 4, 3]
+
+  describe "sumDigits" $
+    it "calculates the sum of sums of the digits of each element in a list" $
+      sumDigits [16, 7, 12, 5] `shouldBe` 22
+
+  describe "validate" $ do
+    it "returns True if an Integer is a valid credit card number" $
```
Comment (flandrade) [inline homework01/test/examples/Main.hs]: @jmorocho What about the following?
```haskell
context "when the credit card number is valid
  it "returns True"
```

File: homework03/test/examples/Main.hs [hunk 23]
```diff
@@ -1,15 +1,39 @@
...
+      skips ([] :: [String]) `shouldBe` ([] :: [[String]])
+      
+  describe "localMaxima" $ do
+    it "finds local maxima and gets 2 elements" $
+      localMaxima [2,9,5,6,1] `shouldBe` [9,6]
+
+    it "finds local maxima and gets 1 elements" $
+      localMaxima [2,3,4,1,5] `shouldBe` [4]
+
+    it "finds local maxima and not gets elements" $
```
Comment (flandrade) [inline homework03/test/examples/Main.hs]: @jmorocho "find local maxima and doesn't get elements"?

In addition, I think we're missing some context functions here. Why in some case do you get one element or two? 

```haskell 
context "list.... "
  it "gets 2 elements"
```

File: homework04/test/examples/Main.hs [hunk 24]
```diff
@@ -1,15 +1,41 @@
...
+    -- Because you can have many implementations of foldTree, you will only test the height of the tree
+    it "returns heigh of a balanced binary tree" $ do      
+      let height (Node x _ _ _) = x
+      height (foldTree "ABCDEFGHIJ") `shouldBe` 3
+         
+  describe "xor" $ do
+    it "returns xor from a list with an odd number of True values" $
+      xor [False, True, False] `shouldBe` True
+   
+    it "returns xor from a list with an even number of True values" $
```
Comment (flandrade) [inline homework04/test/examples/Main.hs]: @jmorocho context `when list has an even number of elements`?

File: homework05/test/examples/Main.hs [hunk 25]
```diff
@@ -1,15 +1,63 @@
...
-    , "src/Calc.hs"
-    ]
+main = hspec $ do
+  
+  describe "eval" $
+    it "returns result of applying add and multiply operations" $
+      eval (ExprT.Mul (ExprT.Add (ExprT.Lit 2) (ExprT.Lit 3)) (ExprT.Lit (-4))) `shouldBe` (-20)
+
+  describe "evalStr" $ do
+    it "evaluates expression with parenthesis" $
```
Comment (flandrade) [inline homework05/test/examples/Main.hs]: @jmorocho What about the following? 

```haskell
context "when an expression has parenthesis"  $ do
  it "evaluates the expression"
```

File: homework05/test/examples/Main.hs [hunk 26]
```diff
@@ -1,15 +1,63 @@
...
+    it "evaluates wrong expression" $
+      evalStr "2+3*" `shouldBe` Nothing
+      
+  describe "reify" $
+    it "makes an instance of Expr to ExprT type" $
+      reify (mul (add (lit 2) (lit 3)) (lit (-4))) 
+        `shouldBe` ExprT.Mul (ExprT.Add (ExprT.Lit 2) (ExprT.Lit 3)) (ExprT.Lit (-4))
+  
+  describe "parseExp" $
+    context "when provided with multiple data types" $ do
```
Comment (flandrade) [inline homework05/test/examples/Main.hs]: @jmorocho I think it's better to have multiple contexts here. For example: 

```haskell
context "when provided with an Integer data type"  $ do
  it "evaluates the expression"
```

What do you think?

File: homework05/test/examples/Main.hs [hunk 27]
```diff
@@ -1,15 +1,63 @@
...
+      it "returns value of the expression evaluated as Mod7" $ do
+        let testSat = parseExp lit add mul "(3 * 4) + (-5)" :: Maybe Mod7
+        testSat `shouldBe` Just (Mod7 0)
+   
+  describe "compile" $
+    it "returns stack operations" $
+      compile "(3*4)" `shouldBe` ((Just [PushI 3,PushI 4,StackVM.Mul]) :: Maybe Program)
+  
+  describe "withVars" $ do
+    it "evaluates all expression if variable exist in a operation of Expr class" $
```
Comment (flandrade) [inline homework05/test/examples/Main.hs]: @jmorocho What about the following? 

```haskell 
context "when the provided variable exists in an operation of Expr class" $ do
  it "evaluates the expression"
```

File: homework01/test/examples/Main.hs [hunk 28]
```diff
@@ -1,15 +1,57 @@
...
+
+      it "returns an empty list for negative inputs" $
+        toDigits (-17) `shouldBe` []
+
+  describe "toDigitsRev" $ 
+    context "when digits are positive integers" $ do
+      it "converts to a list of digits reversed" $ 
+        toDigitsRev 9876 `shouldBe` [6, 7, 8, 9]
+
+      it "returns an empty list for 0" $
```
Comment (flandrade) [inline homework01/test/examples/Main.hs]: @jmorocho I think we're missing some context functions. Perhaps: 

```haskell
context "when the digit is 0"
  it "returns an empty list"
```

File: homework01/test/examples/Main.hs [hunk 29]
```diff
@@ -1,15 +1,57 @@
...
+        toDigitsRev 0 `shouldBe` []
+
+      it "returns an empty list for negative inputs" $
+        toDigitsRev (-1) `shouldBe` []
+
+  describe "doubleEveryOther" $ do
+    it "doubles every other number beginning from the right (even number of elements)" $
```
Comment (flandrade) [inline homework01/test/examples/Main.hs]: @jmorocho Perhaps: 

```haskell
context "when the list has an even number of elements"
  it "doubles every other number beginning from the right"  
```

Could you please verify for each case of `doubleEveryOther`?

**Review by flandrade — COMMENTED** [review]
@jmorocho Good job so far. I left some comments regarding the use of `context`. This might be helpful: http://lmws.net/describe-vs-context-in-rspec

File: homework02/test/examples/Main.hs (hunk 5 above)
Comment (ghost) [reply homework02/test/examples/Main.hs]: Solved :+1:

File: homework04/test/examples/Main.hs (hunk 10 above)
Comment (ghost) [reply homework04/test/examples/Main.hs]: Solved :+1:

File: homework05/test/examples/Main.hs (hunk 13 above)
Comment (ghost) [reply homework05/test/examples/Main.hs]: Solved :+1:

File: homework06/test/examples/Main.hs (hunk 16 above)
Comment (ghost) [reply homework06/test/examples/Main.hs]: Solved :+1:

File: homework01/test/examples/Main.hs (hunk 28 above)
Comment (ghost) [reply homework01/test/examples/Main.hs]: Yes, is a good suggestion.

File: homework01/test/examples/Main.hs (hunk 29 above)
Comment (ghost) [reply homework01/test/examples/Main.hs]: Done.

File: homework01/test/examples/Main.hs (hunk 22 above)
Comment (ghost) [reply homework01/test/examples/Main.hs]: I changed the context in case of valid an invalid credit card number.

File: homework03/test/examples/Main.hs (hunk 23 above)
Comment (ghost) [reply homework03/test/examples/Main.hs]: Context added in case that exists at least one local maximum or none.

File: homework04/test/examples/Main.hs (hunk 24 above)
Comment (ghost) [reply homework04/test/examples/Main.hs]: Context added.

File: homework05/test/examples/Main.hs (hunk 26 above)
Comment (ghost) [reply homework05/test/examples/Main.hs]: Your are right, this test have multiple contexts (Integer, Bool, etc.).

File: homework05/test/examples/Main.hs (hunk 27 above)
Comment (ghost) [reply homework05/test/examples/Main.hs]: Good suggestion, thanks :+1:

File: homework05/test/examples/Main.hs (hunk 25 above)
Comment (ghost) [reply homework05/test/examples/Main.hs]: Added two nested contexts because when the expression is well-formed (with or without parentheses), the function return value of the expression and when the expression is not well-formed (with or without parentheses), return `Nothing`.
